graph-kiota/graph-kiota.md

```diff
@@ -0,0 +1,126 @@
+
+- OpenAPI v1 specification:https://raw.githubusercontent.com/microsoftgraph/msgraph-metadata/master/openapi/v1.0/openapi.yaml
+- OpenAPI beta specification:https://raw.githubusercontent.com/microsoftgraph/msgraph-metadata/master/openapi/beta/openapi.yaml
+
+### Graph libraries using Kiota
+
+#### Graph Service library:
+
+The Graph Service libraries which contain the request builders will be generated using Kiota
+
+Generation command:
+```
+kiota.exe -d https://raw.githubusercontent.com/microsoftgraph/msgraph-metadata/master/openapi/v1.0/openapi.yaml --language typescript -o path\msgraph-sdk-typescript\src -c GraphServiceClient -n MicrosoftGraph
+```
+- GraphServiceClient generated.
+- Request builders generated
+
+**Graph Core library**:
+
+ - Graph core library takes a dependency on Kiota core.
+ - Kiota core contains code for a generic HTTP API. That is:
+    - code to call make the native HTTP client call to the API
+    - any generic middleware.
+ - Graph specific code stays in Graph core. Example: Graph Tasks.
+
+#### Design
+
+
+Graph Core
+
+// Following is for typescript only
+``` typescript
+GraphCoreClient {
+    httpClient:  KiotaHttpClient // importing the HTTPClient from Kiota core library
+}
+```
+
+GraphServiceClient
+```
+class GraphServiceClient{
+    httpCoreInstance : HttpCore // From Kiota
+
+    UserRequestBuilder();
+    MailRequestBuilder();
+}
+
+```
+
+#### Graph core library specifications
+
+This library contains the
+- Factory method/classes returning the:
+    - Core client with the user configurations such as authprovider, custom middleware, request options.
+    - HttpCore instance for the service library with user configurations such as authprovider, custom middleware, request options.
+
+- Tasks
+    - PageIteratorTask
+    - LargeFileUploadTask
+    - Batching
+
+- Authentication Providers
+    - Use the `IAuthProvider` and  auth provider class from the Kiota Auth abstractions.
+
+
+#### Graph Service library specifications
+
+- Discuss about ways to add custom code or custom configurations to the generated `GraphServiceClient` code.
+- The custom configurations that can go here are as follows: 
+ - Setting the default for `HttpCore` with a Graph specific `HttpCore` instance. 
+ - Adding multiple constructors to the `GraphServiceClient`. 
+
+#### Graph Middlewares
+
+- Import middlewares from Kiota core library
+- How to add graph specific checks in the middleware. Example: The JS SDK checks if the url is a Graph url or a custom host passed by the user before appending telemetry headers or auth headers. If not, the headers are removed.
+
+- AuthHandler - Kiota core
+- RetryHandler - Kiota core
+- RedirectHandler - Kiota core
+- CompressionHandler - Kiota core
+- CachingHandler - Kiota core
+- ChaosHandler - Graph core 
+#### Use cases
+
+- Graph core client 
+```
+var response = new coreClient.requestURL("/me").get();
+```
+
+- Graph service client
+```
+var response = new serviceClient.me().get();
```

Comment:
```suggestion
var response = await serviceClient.me.get();
```

I believe we're generating getters for typescript for navigation like that. I'd have to double check